### PR TITLE
Feature/remove transactions queue 1027

### LINKED ISSUE #1027: Get rid of transactions queue
# Problem

Because we don't have "sign later" functionality anymore, there is no reason to have transactions queue on go side.

# Implementation
-remove transactions queue functionality
-instead two `approve` methods implement one `sendTransaction`

image

### PR DESCRIPTION
**Pull request by sebastiandelgado**
Remove `PendingSignRequests` queue from the sign module.

This closes #1027 by removing the pending sign requests queue  dependency from the `SendTransaction`, `SignMessage` and `Recover`. It was split into 3 different PRs that were reviewed before merging into this feature branch:
 - Remove `pendingSignRequests` queue from the transactor and sign transactions without the queue on `SendTransaction` calls
 - Remove `pendingSignRequests` queue from the `personalAPI` service module so that messages are signed/recovered without it in the `SignMessage` and `Recover` methods
 - Remove all of the `pendingSignRequests` code, along with its references in `api/backend.go` and any tests that used it.

The `sign` module was left pretty barebones, so we might want to simplify that in a separate PR (this one is definitely big enough already) by removing most interfaces from them.

@mandrigin @adambabik Thanks for all the help reviewing these PRs. Let me know if you spot anything here before merging.

@flexsurfer please take a look at the `library.go` file which has all the changes to the rpc interface (basically add `SendTransaction`, `SignMessage` and `Recover` methods in place of the `ApproveSignRequest...` methods).

What considerations do I need to take when merging this? For example, if I need to hold off on merging this so you can update the status-react interface to match these changes. What do we need to coordinate with status-react so that we don't break anything when deploying this change?

### CODE REVIEW

**Review by mandrigin — APPROVED** [review]
LGTM

File: api/backend.go
Comment (divan) [inline api/backend.go]: Why is this method not supported? Is it unsupported both with using infura and les?

File: services/personal/api.go
Comment (divan) [inline services/personal/api.go]: It looks like the better approach would be implementing [json.Unmarshaler](https://golang.org/pkg/encoding/json/#Unmarshaler) interface for these types (`SignParams` and `RecoverParams`). So the simple `json.Unmarshal()` will do the job.

File: services/personal/api.go
Comment (divan) [inline services/personal/api.go]: Same as previous comment.

File: transactions/types.go
Comment (divan) [inline transactions/types.go]: This is just a wrapper that runs `json.Unmarshal` and is used only once in `lib/library.go`. I would keep it on the caller side (i.e. in `lib/` instead).

File: api/backend.go
Comment (adambabik) [reply api/backend.go]: Because we want to guard calling it with a password. It's not supported to be called directly, only via a binding. Maybe we can add a comment explaining that.

File: lib/library.go
Comment (adambabik) [inline lib/library.go]: This line should be replaced with `//export SignMessage`. Otherwise, this function won't be public.

File: lib/library.go
Comment (adambabik) [inline lib/library.go]: The same here `//export Recover`.

File: lib/library.go
Comment (adambabik) [inline lib/library.go]: And here `//export SendTransaction`.

File: lib/types.go
Comment (adambabik) [inline lib/types.go]: I run `make lint` locally without this line and there were no errors.

**Review by adambabik — CHANGES_REQUESTED** [review]

**Comment by sebastiandelgado** [comment]
@adambabik @divan Addressed all comments. Only one I addressed differently to the suggestions was the unmarshalling, where I got rid of the unmarshalling methods altogether and simply did the unmarshalling in the lib methods (`SendTransaction`, `Recover`, `SignMessage`)

**Review by adambabik — APPROVED** [review]
Great job! 👏

**Comment by mandrigin** [comment]
@sebastiandelgado added a patch to Makefile to avoid some dep-related leftovers on Jenkins.

**Comment by sebastiandelgado** [comment]
This PR is rebased and ready to merge. Just waiting on confirmation that the interface works as expected before merging into develop

**Comment by adambabik** [comment]
@sebastiandelgado can you please link a status-react PR with this change?

**Comment by flexsurfer** [comment]
@adambabik there is no PR yet, i'm in progress, just started work

**Comment by flexsurfer** [comment]
hey @sebastiandelgado `SignMessage` requires password in rpcParams, and ignores it in function params, also i need params to return to js callback with error and result, also because we don't have failed signal anymore we need this information here about error not only message but id

**Comment by adambabik** [comment]
What left here is to return a specific error code for `keystore.ErrDecrypt`.

**Comment by adambabik** [comment]
@flexsurfer Codes for `eth_sendTransaction`, `personal_sign` and `personal_ecRecover`:
```
// special codes
codeFailedParseResponse = 1
codeFailedParseParams = 2
// account related codes
codeErrNoAccountSelected = 3
codeErrInvalidTxSender = 4
codeErrDecrypt = 5
```

**Comment by flexsurfer** [comment]
https://docs.google.com/drawings/d/1l472xaRgEJzizxUkNmeL0STSNqX9ClUy0OoZpKEjjXo

new TX flow

**Comment by flexsurfer** [comment]
@adambabik before it was not possible to send the same tx twice because we had a queue on go side, do you have some protection in this PR from sending the same tx twice? for example, if we call `SendTransaction` twice you can't distinguish that this is the same tx and will send it twice?

**Comment by adambabik** [comment]
@flexsurfer how is it possible to send two transactions? I though that after this change each `eth_sendTransaction` call will result in a blocking confirmation screen. Until it's canceled or confirmed it won't be possible to send another transaction. Is that not true?

From status-go point of view, each call of `SendTransaction` binding is a separate transaction, there is nothing to protect from sending the same tx twice (they will get different nonces) and there wasn't any protection before. With the queue mechanism, two transactions would be scheduled.

**Comment by flexsurfer** [comment]
@adambabik we had a bug in the past, when it was possible to press sign button twice and we called approveTransaction twice, but we had tx id (because you had a queue on go side), nothing wrong happened , but now if we'll have same bug in status-react two transactions will be sent, so we need to be super careful on status-react side now

**Review by flexsurfer — APPROVED** [review]
tested, ready to be merged!

**Comment by adambabik** [comment]
@flexsurfer success! Check out https://ci.status.im/job/status-go/job/status-go-manual/302/consoleFull. I built it from a tag, can you check if `tags-v0.11.0^0-g4afd9e6c-302` is gonna work? It has a strange artefact :/

<blockquote><div><strong><a href="https://ci.status.im/job/status-go/job/status-go-manual/302/consoleFull">status-go » status-go-manual #302 Console [Jenkins]</a></strong></div></blockquote>

**Pull request merged**
